src/main/java/com/epam/learn/springcore/dao/TrainingRepository.java

```diff
@@ -1,14 +1,23 @@
 package com.epam.learn.springcore.dao;
 
-import com.epam.learn.springcore.entity.Trainee;
-import com.epam.learn.springcore.entity.Trainer;
 import com.epam.learn.springcore.entity.Training;
+import com.epam.learn.springcore.specification.TraineeTrainingSearchCriteria;
+import com.epam.learn.springcore.specification.TraineeTrainingSpecification;
+import com.epam.learn.springcore.specification.TrainerTrainingSearchCriteria;
+import com.epam.learn.springcore.specification.TrainerTrainingSpecification;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
 
 import java.util.List;
 
-public interface TrainingRepository extends JpaRepository<Training, Integer> {
-//    List<Trainer> findUnassignedTrainers(String traineeUsername);
-//    void updateTrainersListForTrainee(String traineeUsername, List<Trainer> trainers);
-    Training findByTraineeAndTrainer(Trainee trainee, Trainer trainer);
+public interface TrainingRepository extends JpaRepository<Training, Integer>, JpaSpecificationExecutor<Training> {
+
+    default List<Training> findByCriteria(TraineeTrainingSearchCriteria criteria) {
```

Comment:
why did you use default?

Reply:
It is not good to send one entity and then move it to another inside repository method

Reply:
Updated